app/validation/validator.py

```diff
@@ -559,18 +579,27 @@ def _get_numeric_range_values(self, answer, answer_ranges):
         }
 
     def validate_duplicates(self, json_to_validate):
+        """
+        question_id & answer_id should be globally unique with some exceptions:
+            - within a block, ids can be duplicated across variants, but must still be unique outside of the block.
+        """
+
         special_keys = ['id']
         duplicate_errors = []
 
         for special_key in special_keys:
-            unique_items = []
+            items_per_block = defaultdict(set)
 
-            for value in self._parse_values(json_to_validate, special_key):
-                if value in unique_items:
-                    duplicate_errors.append(
-                        self._error_message('Duplicate {} found. value {}'.format(special_key, value)))
-                else:
-                    unique_items.append(value)
+            for path, value in self._parse_values(json_to_validate, special_key):
+                if 'blocks' in path:
+                    block_path = path[:path.index('blocks') + 2]
+                    string_path = '.'.join(block_path)
+                    items_per_block[string_path].add(value)
```

Comment:
Maybe worth a comment around this bit of code, took me a while to work out how it was working

Reply:
Agreed, it's not immediately obvious. Maybe there's a simpler way to do it?

Reply:
I like the approach, using the uniqueness of a set, just maybe a comment to highlight thats what it is doing